baybe/searchspace/discrete.py

```diff
@@ -298,52 +308,113 @@ def from_simplex(
-        # way. More specifically: after having cross-joined a new parameter, there must
-        # be enough "room" left for the remaining parameters to fit. Hence,
-        # configurations of the current parameter subset that exceed the desired
-        # total value minus the minimum contribution to come from the yet to be added
-        # parameters can be already discarded.
-        for i, (param, min_to_go) in enumerate(
-            zip_longest(simplex_parameters, min_upcoming, fillvalue=0)
+        # way. More specifically:
+        # * After having cross-joined a new parameter, there must
+        #   be enough "room" left for the remaining parameters to fit. That is,
+        #   configurations of the current parameter subset that exceed the desired
+        #   total value minus the minimum contribution to come from the yet-to-be-added
+        #   parameters can be already discarded, because it is already clear that
+        #   the total sum will be exceeded once all joins are completed.
+        # * Analogously, there must be enough "nonzero slots" left for the yet to be
+        #   joined parameters, i.e. parameter subset configurations can be discarded
+        #   where the number of nonzero parameters already exceeds the maximum number
+        #   of nonzeros minus the number of nonzeros to come, because it is already
+        #   clear that the maximum will be exceeded once all joins are completed.
+        # * Similarly, it can be verified for each parameter that there are still
+        #   enough nonzero parameters to come to even reach the minimum
+        #   desired number of nonzero after all joins.
+        for i, (
+            param,
+            min_sum_to_go,
+            min_nonzero_to_go,
+            max_nonzero_to_go,
+        ) in enumerate(
+            zip_longest(
+                simplex_parameters,
+                min_sum_upcoming,
+                min_nonzero_upcoming,
+                max_nonzero_upcoming,
+                fillvalue=0,
+            )
         ):
             if i == 0:
                 exp_rep = pd.DataFrame({param.name: param.values})
             else:
                 exp_rep = pd.merge(
                     exp_rep, pd.DataFrame({param.name: param.values}), how="cross"
                 )
-            drop_invalid(exp_rep, max_sum - min_to_go, boundary_only=False)
+            drop_invalid(
+                exp_rep,
+                max_sum=max_sum - min_sum_to_go,
+                # the maximum possible number of nonzeros to come dictates if we
+                # can achieve our minimum constraint in the end:
+                min_nonzero=min_nonzero - max_nonzero_to_go,
+                # the minimum possible number of nonzeros to come dictates if we
+                # can stay below the targeted maximum in the end:
+                max_nonzero=max_nonzero - min_nonzero_to_go,
+                boundary_only=False,
+            )
```

Comment:
is there a bug down in line 430?

it says
`cls(parameters=simplex_parameters, exp_rep=exp_rep)`

but shouldnt it be 
`cls(parameters=simplex_parameters+product_parameters, exp_rep=exp_rep)`

Reply:
(if im right and this is a bug the tests dont capture this at the moment)

Reply:
From what I understand here, the `product_parameters` are included in the `exp_rep` (line 424) but not in the construction of the class. So I would agree that this is a bug here.

Reply:
It seems like there is no validation function for the exp_rep, right? I guess we should add some check that verifies that the columns of exp rep and the names of the parameters are identical as a set, right?